src/Build/BuildCheck/Infrastructure/BuildCheckManagerProvider.cs

```diff
@@ -380,12 +391,26 @@ public void RemoveChecksAfterExecutedActions(List<CheckWrapper>? checksToRemove,
 
         private void RemoveCheck(CheckFactoryContext checkToRemove)
         {
-            _checkRegistry.Remove(checkToRemove);
+            var newRegistry = new ConcurrentBag<CheckFactoryContext>(_checkRegistry.Where(x => x != checkToRemove));
+
+            while (_checkRegistry.TryTake(out _)) { }
+
+            foreach (var item in newRegistry)
+            {
+                _checkRegistry.Add(item);
+            }
```

Comment:
```suggestion
            _checkRegistry = newRegistry;
```

Is this acheicvving smae?

Reply:
I was afraid of facing the same situation: if other threads are currently accessing or modifying the `_checkRegistry`, but I might be playing it extra safe

Reply:
I don't think this idea improves the situation. 
T1 RemoveCheck -> create newRegistry
T2 add check C to _checkRegistry
T1 while ... foreach ... _checkregistry=newregistry

outcome: C is not in _checkRegistry :(

why not a concurrentdictionary?

Reply:
Well, I am not sure what Concurrent dictionary is supposed to address here.
It was a switch from not thread safe to a safe alternative. 
Could you please elaborate?

Reply:
The goal is to remove one element, while leaving the collection in a consistent state, right? ConcurrentBag does not have an api for that, so you'd have to use locks I think. ConcurrentDictionary does have the api https://learn.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.tryremove?view=net-9.0#system-collections-concurrent-concurrentdictionary-2-tryremove(-0-1@)

the only drawback is that you have to come up with a key

Reply:
Eah - correct :-/
I'm still slightly leaning towards the ConcurrentBag as we'll be interating order of magnitude more times then modifying. We can guard the adds/removes with a lock and then keep ConcurrentBag - what do you think?

Remove:
```
lock(_someLock)
{
    _checkRegistry = new ConcurrentBag<CheckFactoryContext>(_checkRegistry.Where(x => x != checkToRemove));
}
```

Add:
```
lock(_someLock)
{
    _checkRegistry.Add(...)
}
```

Traverse:
```
  foreach(var check in _checkRegistry)
  {
  }
```

We can have stale reads - but that's fine. We want to prevent overwriten updates and exceptions on iterating

Reply:
Or keep _checkRegistry readonly, but making sure we do not remove items without checking what we're removing

Reply:
I'm not up to date with the performance benefits of Concurrentbag over dictionary in this scenario. If you think that's worth the slight code cumbersomeness, OK.